access/discord/bot.go

```diff
@@ -0,0 +1,183 @@
+const discordRedColor = 13771309  // Green OxD2222D
+const discordGreenColor = 2328611 // Red 0x2328611
+
+// DiscordBot is a discord client that works with AccessRequest.
+// It's responsible for formatting and posting a message on Discord when an
+// action occurs with an access request: a new request popped up, or a
+// request is processed/updated.
+type DiscordBot struct {
+	client      *resty.Client
+	clusterName string
+	webProxyURL *url.URL
+}
+
+// onAfterResponseDiscord resty error function for Discord
+func onAfterResponseDiscord(_ *resty.Client, resp *resty.Response) error {
+	if resp.IsSuccess() {
+		return nil
+	}
+
+	var result DiscordResponse
+	if err := json.Unmarshal(resp.Body(), &result); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if result.Message != "" {
+		return trace.Errorf("%s (code: %v, status: %d)", result.Message, result.Code, resp.StatusCode())
+	}
+
+	return trace.Errorf("Discord API returned error: %s (status: %d)", string(resp.Body()), resp.StatusCode())
+}
+
+func (b DiscordBot) CheckHealth(ctx context.Context) error {
+	_, err := b.client.NewRequest().
+		SetContext(ctx).
+		Get("/users/@me")
+	if err != nil {
+		return trace.Wrap(err, "health check failed, probably invalid token")
+	}
+
+	return nil
+}
+
+// Broadcast posts request info to Discord.
+func (b DiscordBot) Broadcast(ctx context.Context, recipients []common.Recipient, reqID string, reqData pd.AccessRequestData) (common.SentMessages, error) {
+	var data common.SentMessages
+	var errors []error
+
+	for _, recipient := range recipients {
+		var result ChatMsgResponse
+		_, err := b.client.NewRequest().
+			SetContext(ctx).
+			SetBody(DiscordMsg{
+				Msg:  Msg{Channel: recipient.ID},
+				Text: b.discordMsgText(reqID, reqData),
+			}).
+			SetResult(&result).
+			Post("/channels/" + recipient.ID + "/messages")
+		if err != nil {
+			errors = append(errors, trace.Wrap(err))
+			continue
+		}
+		data = append(data, common.MessageData{ChannelID: recipient.ID, MessageID: result.DiscordID})
+
+	}
+
+	return data, trace.NewAggregate(errors...)
+}
+
+// PostReviewReply does nothing as Discord does not have threaded replies
+func (b DiscordBot) PostReviewReply(ctx context.Context, channelID, timestamp string, review types.AccessReview) error {
+	return nil
+}
+
+// UpdateMessages updates already posted Discord messages
+func (b DiscordBot) UpdateMessages(ctx context.Context, reqID string, reqData pd.AccessRequestData, messagingData common.SentMessages, reviews []types.AccessReview) error {
+	var errors []error
+	for _, msg := range messagingData {
+		_, err := b.client.NewRequest().
+			SetContext(ctx).
+			SetBody(DiscordMsg{
+				Msg:    Msg{Channel: msg.ChannelID},
+				Text:   b.discordMsgText(reqID, reqData),
+				Embeds: b.discordEmbeds(reviews),
+			}).
+			Patch("/channels/" + msg.ChannelID + "/messages/" + msg.MessageID)
+		if err != nil {
+			errors = append(errors, trace.Wrap(err))
+		}
+	}
+
+	if len(errors) > 0 {
+		return errors[0]
+	}
```

Comment:
We're losing errors here
Could we print them out? or return all of them using `trace.Aggregate`

Reply:
Done